### PR TITLE
Remove routing information from RootModel and allow multiple ObsIds in relevant URLs

### PR DESCRIPTION
**Pull request by toddburnside**
This PR makes the URL being the source of truth for focused observations and targets. It fixes most, but not all, of the back/forward navigation issues we have been seeing. There are still a few cases where back does not work properly.

In a followup PR I am going to rework how we are using TwoPanelState. Because the focused obs only contained a single id, and we allowed multi-select, we need to track what was actually selected. This can be simplified now. I think this may also fix at least some of the remaining back button issues. But, I wanted to get this PR in because it is working better than current staging, and it has a lot of file changes and I want to avoid merge conflicts.

### CODE REVIEW

File: explore/src/main/scala/explore/Routing.scala [hunk 1]
```diff
@@ -86,77 +90,96 @@ object Routing {
...
       import dsl._
 
+      // We can't use gid.regexPattern because capture groups don't work here.
+      def gidRegEx[Id](implicit gid: Gid[Id]): String =
+        s"${gid.tag.value.toString}-[1-9a-f][0-9a-f]*"
+
       def id[Id](implicit gid: Gid[Id]): StaticDsl.RouteB[Id] =
-        // We can't use gid.regexPattern because capture groups don't work here.
-        string(s"${gid.tag.value.toString}-[1-9a-f][0-9a-f]*").pmapL(gid.fromString)
+        string(gidRegEx).pmapL(gid.fromString)
```
Comment (cquiroz) [inline explore/src/main/scala/explore/Routing.scala:98]: oh what is this? is `string` defined in `scalajs-react`?

File: common/src/main/scala/explore/model/AppContext.scala [hunk 2]
```diff
@@ -105,21 +107,46 @@ case class AppContext[F[_]](
...
   sso:         SSOClient[F],
-  pageUrl:     (AppTab, Option[Observation.Id], Option[Target.Id]) => String,
-  setPage:     (AppTab, Option[Observation.Id], Option[Target.Id]) => Callback,
+  pageUrl:     (AppTab, Option[ObsIdSet], Option[Target.Id]) => String,
+  setPage:     (AppTab, Option[ObsIdSet], Option[Target.Id]) => Callback,
+  setPageVia:  (
+    AppTab,
+    Option[ObsIdSet],
+    Option[Target.Id],
+    SetRouteVia
```
[the remark this answers is not in the document]
Comment (toddburnside) [reply common/src/main/scala/explore/model/AppContext.scala:116]: SetRouteVia is what scalajs-react uses to allow you to push your url onto the history stack, or replace the current item at the top of the stack with the new url. By default we want to push, but there are occasions where we want to replace - for example if the url contains an invalid obs id set. If we were to push, going back would give the bad url, which we would replace again. This currently is happening in staging.

File: explore/src/main/scala/explore/Routing.scala (hunk 1 above)
Comment (toddburnside) [reply explore/src/main/scala/explore/Routing.scala:98]: Yeah. `string` is part of the scalajs-react routing DSL. It essentially extracts a bit of the route based on the regex. You can modify further with the `pMap*` methods.

File: common/src/main/scala/explore/model/AppContext.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (toddburnside) [reply common/src/main/scala/explore/model/AppContext.scala:116]: I thought about creating our own ADT to hide the scalajs-react `SetRouteVia`, but decided that was a bit much.

[the remark this answers is not in the document]
Comment (rpiaggio) [reply common/src/main/scala/explore/model/AppContext.scala:116]: Can we just store the `setPageVia` function, and then define `setPage` as a method with a default value for `SetRouteVia`?

[the remark this answers is not in the document]
Comment (toddburnside) [reply common/src/main/scala/explore/model/AppContext.scala:116]: Yeah. We can do that.

**Comment by rpiaggio** [comment]
Thank you so much for tackling this.

It simplifies the routing and definitely makes it easier to reason about. :)

I'm wondering if we can pass down a `View[RoutingInfo]` down to components such that they are abstracted from the scalajs-react API. It wouldn't be a `View` over regular React state, but we can build the `View` such that calling `mod` on `RoutingInfo` just invokes `setPage`. Components would also be able to just set ie the selected observation by zooming into the `View`.

We can explore this in another PR.

**Review by rpiaggio — APPROVED** [review]
LGTM

File: common/src/main/scala/explore/model/AppContext.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (toddburnside) [reply common/src/main/scala/explore/model/AppContext.scala:115]: Pushed a commit that keeps the `setPageVia` method, but renames everything else to `pushPage*` or `replacePage*`. I think this makes it clearer in the code, and most places don't have to care about `SetRouteVia`.

**Pull request merged**